gpii/node_modules/preferencesServer/test/cloudSafeCredTests.js

```diff
@@ -0,0 +1,375 @@
+/**
+GPII Preferences Server Tests
+
+Copyright 2018 Raising the Floor - International
+
+Licensed under the New BSD license. You may not use this file except in
+compliance with this License.
+
+The research leading to these results has received funding from the European Union's
+Seventh Framework Programme (FP7/2007-2013) under grant agreement no. 289016.
+
+You may obtain a copy of the License at
+https://github.com/GPII/universal/blob/master/LICENSE.txt
+*/
+
+"use strict";
+
+var fluid = require("infusion"),
+    jqUnit = fluid.require("node-jqunit", require, "jqUnit"),
+    gpii = fluid.registerNamespace("gpii");
+
+require("./preferencesServerTestsUtils.js");
+require("./preferencesServerTests.js");
+
+fluid.require("%gpii-universal");
+gpii.loadTestingSupport();
+
+fluid.registerNamespace("gpii.tests.preferencesServer.cloudSafeCred");
+
+gpii.tests.preferencesServer.config = {
+    configName: "gpii.tests.preferencesServer.config",
+    configPath: "%preferencesServer/test/configs"
+};
+
+/////////////// TEST Creating a Cloud Safe Login /////////////////
+//
+// Steps 1: Send the POST to create a cloudSafeCred for an existing prefset
+// Step 2: Verify that the Cloud Safe Cred Document is correct
+// Step 3: Verify that the gpii-express-user document is correct
+// Step 4: Verify that this is included in the list keys endpoint
+// Step 5: Verify that this is included in the prefssafe with keys endpoint
+fluid.registerNamespace("gpii.tests.preferencesServer.cloudSafeCred.put");
+
+/*
+ * Test response payload against a subset of a gpiiCloudSafeCred document.
+ */
+gpii.tests.preferencesServer.cloudSafeCred.testCloudSafeCreatePut = function (cloudCredSubset, response) {
+    var data = JSON.parse(response);
+    jqUnit.assertLeftHand("Returned cred document should have the following: ", cloudCredSubset, data);
+};
+
+/*
+ * Check and see if the prefs safe that came back has our new credentials doc.
+ */
+gpii.tests.preferencesServer.cloudSafeCred.testCloudSafeCred = function (response /*, expected, receivedStatusCode, expectedStatusCode */) {
+    var data = JSON.parse(response);
+    jqUnit.assertEquals("There should be a prefsSafe", data.prefsSafe.type, "prefsSafe");
+    jqUnit.assertEquals("There should be a cloud cred", data.keys[0].type, "gpiiCloudSafeCredential");
+    jqUnit.assertEquals("There should be a gpii key", data.keys[1].type, "gpiiKey");
+};
+
+gpii.tests.preferencesServer.cloudSafeCred.put.buildSuccessTestDef = function (fixture) {
+    var expectedStatusCode = fixture.expectedStatusCode || 200;
+    return {
+        name: fixture.name,
+        config: gpii.tests.preferencesServer.config,
+        components: {
+            putCloudCred: {
+                type: "kettle.test.request.http",
+                options: {
+                    path: "/add-cloud-credentials/" + fixture.url,
```

Comment:
Using the same steps described [here](https://github.com/GPII/universal/pull/634/#discussion_r266045546) to send PUT request to `http://localhost:8081/add-cloud-credentials/prefsSafe-carla` to reproduce these issues. Please write a test to cover these cases too:

1. When not providing PUT body or providing PUT body that does not have "password" field in the object, GPII terminates with this error:
```
FATAL ERROR: Uncaught exception: Pass phrase must be a buffer
TypeError: Pass phrase must be a buffer
    at pbkdf2 (crypto.js:702:20)
    at Object.exports.pbkdf2Sync (crypto.js:687:10)
...
```

2. Add credential for a prefs safe that never has a credential record by providing PUT body with "password" without "username" such as:
```
{
	"password": "bbb"
}
```
Receives this reponse:
```
{
    "isError": true,
    "message": "Unable to create credentials with that username and password combination"
}
```
Meanwhile, GPII console log shows:
```
Error adding cloud safe credential{
    "isError": true,
    "message": {
        "error": "conflict",
        "reason": "Document update conflict."
    }
}
```
`Document update conflict` error is odd because there has never been a credential record created for this prefs safe.

Reply:
If the given username is an empty string or a couple of spaces, `PUT /add-cloud-credentials` endpoint will still go ahead to create these usernames. There should be stronger sanitization.

Reply:
I've added unit tests and validation to make sure the payload is basically sane, has both fields and they aren't empty.  There is somewhat extensive rules being developed by UX and requirements for what should be considered allowable for deployments, and would like to leave the work for those more complex rules to another jira once the requirements are locked down.